[DOC-307] [2024.1] Replace/decommision a node (#22207)
yugabyte/yugabyte-db: docs/content/preview/yugabyte-platform/manage-deployments/remove-nodes.md

type: docs
---

## Replace a live or unreachable node

> If a node (live or unreachable) needs to be replaced, do the following:

Reviewer: ```suggestion
To replace a live node for extended maintenance or replace an unhealthy node, do the following: 
```

1. Navigate to **Universes**, select your universe, and open the **Nodes** tab.
1. Click the associated node **Actions > Replace Node**.

    ![Replace Node Actions](/images/ee/replace-node.png)

1. Click OK to confirm.

YugabyteDB Anywhere (YBA) starts the node replacement process, and you can view the progress on the **Tasks** tab. As part of the node replacement process, all data (tablets) on the existing node will be moved to other nodes to ensure that the desired replication factor is maintained throughout the operation.

> For cloud providers (AWS, Azure, or GCP), YBA returns the node back to the provider.

Reviewer: ```suggestion
For cloud providers (AWS, Azure, or GCP), YBA returns the existing node back to the provider and provisions a new replacement node from the cloud provider. For on-premises universes, the existing node is returned to the [on-premises provider node pool](../configure-yugabyte-platform/on-premises-nodes/) and a new replacement node is selected from the free pool.
```

> For on-premises universes, because the unresponsive node will likely fail to clean up its data directory, home directory, and so on, YBA sets the state to Decommissioned. This prevents the node from being added to a new universe.

Reviewer: ```suggestion
For on-premises universes, clean up of existing data directories and running processes may fail if the node is unhealthy. In such cases, YBA sets the state to Decommissioned. This prevents the node from being added to a new universe.
```

### Check on-premises node state

On-premises nodes have three states: In use, Free, and Decommissioned as described in the following illustration.

![Decommissioned node workflow](/images/ee/on-prem-replace-workflow.png)
To check the state of an on-premises node, navigate to **Configs > Infrastructure > On-Premises Datacenters**, select the associated on-premises configuration, and click **Instances**.

### Recommission a decommissioned on-premises node

You can return a Decommissioned node to the on-premises provider Free pool after it becomes reachable and is cleaned up.

Perform the following steps to recommission a node:

1. Navigate to **Configs > Infrastructure > On-Premises Datacenters**, select the associated on-premises configuration, and click **Instances**.

> 1. Under Instances, for the decommissioned node, click **Actions > Recommission Node**.

Reviewer: ```suggestion
1. Under Instances, for the decommissioned node, click **Actions > Recommission Node**. YBA will now re-attempt to clean up existing data directories and processes on this node.
```

    ![Recommission Node](/images/ee/recommission-node.png)

1. Click OK to confirm.

YugabyteDB Anywhere (YBA) starts the node recommissioning process, and you can view the progress on the **Tasks** tab.

## Eliminate an unresponsive node

If a virtual machine or a physical server in a universe reaches its end of life and has unrecoverable hardware or other system issues (such as problems with its operating system, disk, and so on) it is detected and displayed in the YugabyteDB Anywhere UI as an unreachable node, as per the following illustration:

![Unreachable Node Actions](/images/ee/node-actions-unreachable.png)

> When this happens, new Master leaders are elected for the underlying data shards. Because the universe enters a partially under-replicated state, it would not be able to tolerate additional failures. To remedy the situation, it is recommended that you [replace the node](#replace-a-live-or-unreachable-node).

Reviewer: ```suggestion
When this happens, new tablet leaders are elected for the underlying data shards and availability is maintained. If no other nodes are available to rebuild the lost copies of these tablets, the universe enters a partially under-replicated state and it would not be able to tolerate additional failures. To remedy the situation, it is recommended that you [replace the node](#replace-a-live-or-unreachable-node).
```

type: docs
---

## Replace a live or unreachable node

If a node (live or unreachable) needs to be replaced, do the following:

1. Navigate to **Universes**, select your universe, and open the **Nodes** tab.
1. Click the associated node **Actions > Replace Node**.

    ![Replace Node Actions](/images/ee/replace-node.png)

> Clicking the **Replace Node** button opens a popup window to confirm the replacement of this node. Confirming the task redirects to the task page and will take a few minutes to complete.

Reviewer 2: I don't think confirming replacing redirects to task page, it should reload the page and the loading circle and the progress bar should slowly run

@rajmaddy89  can you confirm?

Author: I have just added this instead and I think this should suffice - 
1. Click OK to confirm.
YugabyteDB Anywhere (YBA) starts the node replacement process, and you can view the progress on the **Tasks** tab.

For Infrastructure as a service (IaaS) such as AWS and GCP, YugabyteDB Anywhere returns the node back to the IaaS. For on-premises universes, the unresponsive node will most likely have failed to clean up its data directory, home directory, and more. So it will be placed in a Decommissioned state in the on-premises provider which prevents the node from being added to a new universe.

### Check on-premises node state

The on-premises node instances have three states: In use, Free, and Decommissioned and the following illustration describes the workflow of the states.

> ![Decommissioned node workflow](/images/ee/decommissioned-node.png)

Reviewer 2: We can remove the "Action: Decommission", as we do not support this currently

Reviewer 2: Also instead of release node in the "Release node: data cleanup fails" and "Release node:  cleanup succeeds" let's change "Release node" to "Replace node"

> To check the state of an on-premises node, navigate to **Configs > On-Premises Datacenters**, then click its associated on-premises configuration, and click **Instances**.

Reviewer 2: Would be good to have a picture for this

cc @rajmaddy89

### Recommission on-premises node instance

A node in "Decommissioned" state can be placed back into the "Free" node state pool after it becomes reachable and is cleaned up.

Perform the following steps to recommission a node.

> 1. For a node instance in the Decommissioned state, navigate to **Configs** > On-Premises Datacenters**, then click its associated on-premises configuration, and click **Instances**.

Reviewer 2: After "On-Premises Datacenters" UI shows "**" which is incorrect

> 1. On the Instances page, click **Actions > Recommission Node**.

Reviewer 3: ```suggestion
1. Under Instances, for the decommissioned node, click **Actions > Recommission Node**.
```

    ![Recommission Node](/images/ee/recommission-node.png)

> Clicking the **Recommission Node** button opens a popup window to confirm the recommission of this node. Confirming the task redirects to the task page and will take a few minutes to complete.

Reviewer 3: ```suggestion
1. Click OK to confirm.

YugabyteDB Anywhere starts the node recommissioning process, and you can view the progress on the **Tasks** tab.
```

> ## Eliminate an unresponsive node

Reviewer 2: We should say now that replace is the recommended flow for elinating an unreachable node. 

And that now remove, release, delete, add are not recommended and are put in the advanced actions

Reviewer 2: Will need pics for this

cc @rajmaddy89

Reviewer 2: In this case, we can put the start/stop node above remove, release, add, delete, as it will be used more often

Author: @charleswang234 - If its not recommended, then should those sequence of steps just be removed? For what situations would a user actually choose it ? 

At the moment ill comment it out and recommend using replace.

Reviewer 2: Currently we have remove, release, delete, add for old users, as they are used to that workflow. Other than that, it should not be used.

{{< /note >}}

## Start and stop node processes

### Stop a process

> If a node needs the intervention, you can click its associated **Actions > Stop Processes**.

Reviewer: ```suggestion
If a node needs to be briefly taken out of service (for example, to perform a quick OS patch), you can click its associated **Actions > Stop Processes**. It is expected that this node will be returned to service soon through the **Actions > Start Processes** operation.
```

After the YB-TServer and (where applicable) YB-Master server are stopped, the node status is updated and the instance is ready for the planned system changes.

Generally, when a YB-Master is stopped on a node, YugabyteDB Anywhere automatically attempts to start a new YB-Master on another node in the same Availability Zone as the node on which YB-Master is stopped. This ensures that the number of YB-Master servers equals the replication factor (RF) and YB-Master servers are never underreplicated.

> It is recommended not to stop more than (RF - 1) / 2 processes at any given time. For example, on an RF=3 cluster with three nodes, there can only be one node with stopped processes to allow the majority of the nodes to perform Raft consensus operations.

Reviewer: @charleswang234 : Should we rephrase this to say we don't recommend stopping multiple nodes at the same time because they can contain copies of the same tablet?

Reviewer 2: Yes, we should recommend that we should not stop multiple nodes at the same time (especially if in the same az).

> It is recommended not to stop more than (RF - 1) / 2 processes at any given time. For example, on an RF=3 cluster with three nodes, there can only be one node with stopped processes to allow the majority of the nodes to perform Raft consensus operations.

Reviewer: ```suggestion
In general, it is not recommended to execute the "Stop Processes" action on multiple nodes simultaneously. For example, two stopped nodes might share a common tablet. This could cause unavailability on a universe with replication factor 3.
```

### Start a process

You can restart the node's processes by navigating to **Universes**, selecting your universe, then selecting **Nodes**, and then clicking **Actions > Start Processes** corresponding to the node. This returns the node to the Live state.

> In some cases, the system might experience an unrecoverable error. To mitigate, you can use the **Release Instance** option for the stopped node. This also removes the backing instance. For details, see [Remove a node](#remove-node).

Reviewer: @charleswang234 : In the context of start process, we don't routinely expect unrecoverable errors right? Do we need this part?

Reviewer 2: I don't think we need line 97 anymore, since now we only allow retries on the same task for node ops

> In some cases, the system might experience an unrecoverable error. To mitigate, you can use the **Release Instance** option for the stopped node. This also removes the backing instance. For details, see [Remove a node](#remove-node).

Reviewer: ```suggestion
```
